Zend/zend_string.c

```diff
@@ -148,22 +148,7 @@ static zend_always_inline zend_string *zend_interned_string_ht_lookup_ex(zend_ul
 
 static zend_always_inline zend_string *zend_interned_string_ht_lookup(zend_string *str, HashTable *interned_strings)
 {
-	zend_ulong h = ZSTR_H(str);
-	uint32_t nIndex;
-	uint32_t idx;
-	Bucket *p;
-
-	nIndex = h | interned_strings->nTableMask;
-	idx = HT_HASH(interned_strings, nIndex);
-	while (idx != HT_INVALID_IDX) {
-		p = HT_HASH_TO_BUCKET(interned_strings, idx);
-		if ((p->h == h) && zend_string_equal_content(p->key, str)) {
-			return p->key;
-		}
-		idx = Z_NEXT(p->val);
-	}
-
-	return NULL;
+	return zend_interned_string_ht_lookup_ex(ZSTR_H(str), ZSTR_VAL(str), ZSTR_LEN(str), interned_strings);
```

Comment:
@nikic FYI - I wonder if you're able to reproduce the test failure in the prior commit (in gcc `9.3.0-17ubuntu1~20.04`) for `make test TESTS='ext/spl/tests/Vector/aggregate.phpt -m --show-mem --show-diff` (it segfaults in `idx = Z_NEXT(p->val);` when compiling the code and interning the strings, not when running it). It seems almost definitely related to running the code under valgrind with `-O2 [-g]`. Even after `git clean -fdx` and recompiling it's still an issue

- Oddly, running valgrind tests in docker in other OSes (e.g. dockerhub php:8.1.0RC1 with `gcc (Debian 10.2.1-6) 10.2.1 20210110`) don't have this issue.
- When I add `fprintf(stderr` statements, the issue goes away

(The segfault doesn't happen with gdb, or even gdb with `USE_ZEND_ALLOC=0` - it's baffling.)

- If you are able to reproduce this, it's probably a symptom of a larger issue that may affect other php 8.1/8.2 users using the same compiler
- If you aren't, it still may be a good idea to avoid (1) repeating code and (2) redundant loads of ZSTR_LEN(str) from memory into registers if gcc can't infer that the assembly function zend_string_equal_val doesn't modify memory in the assembly block, and needs to scan over multiple buckets